presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,739 @@
+            return;
+        }
+
+        outputBuffers.checkValidTransition(newOutputBuffers);
+    }
+
+    @Override
+    public synchronized ListenableFuture<BufferResult> get(OutputBuffers.OutputBufferId bufferId, long token, DataSize maxSize)
+    {
+        requireNonNull(bufferId, "outputBufferId is null");
+        checkArgument(bufferId.getId() == 0, "Invalid buffer id");
+        checkArgument(maxSize.toBytes() > 0, "maxSize must be at least 1 byte");
+
+        long currentSequenceId;
+
+        acknowledge(bufferId, token);
+
+        currentSequenceId = sequenceId.get();
+
+        // process the request if we have no more data coming in, have data to read, or if this is an outdated request
+        if (noMorePages.get() || !handleInfoQueue.isEmpty() || !pages.isEmpty() || currentSequenceId != token) {
+
+            return processRead(token, maxSize);
+        }
+
+        // if we have no pending read, create one. otherwise, return the future from ongoing pending read
+        if (pendingRead == null) {
+            pendingRead = new PendingRead(taskInstanceId, currentSequenceId, maxSize);
+        }
+
+        return pendingRead.getResultFuture();
+    }
+
+    private void processPendingRead(PendingRead pendingRead)
+    {
+        if (pendingRead.getResultFuture().isDone()) {
+            return;
+        }
+
+        ListenableFuture<BufferResult> resultFuture = processRead(pendingRead.getSequenceId(), pendingRead.getMaxSize());
+        pendingRead.setResultFuture(resultFuture);
+    }
+
+    private synchronized ListenableFuture<BufferResult> processRead(long token, DataSize maxSize)
+    {
+        long currentSequenceId;
+
+        currentSequenceId = sequenceId.get();
+
+        // token is for a page before the current page position
+        if (token < currentSequenceId) {
+            return immediateFuture(emptyResults(taskInstanceId, token, false));
+        }
+
+        // tells client that buffer is complete
+        if (noMorePages.get() && handleInfoQueue.isEmpty() && pages.isEmpty()) {
+            return immediateFuture(emptyResults(taskInstanceId, token, true));
+        }
+
+        // validate previous pages were acknowledged
+        checkState(currentSequenceId == token, "Invalid token");
+
+        // get a copy of in memory pages and the HandleInfo to save the handleFuture
+        List<HandleInfo> handleInfos = ImmutableList.copyOf(handleInfoQueue);
+        List<SerializedPage> pages = ImmutableList.copyOf(this.pages);
+
+        GetTracker getTracker = new GetTracker(maxSize, handleInfos, pages, toIntExact(startPage.get()));
+
+        return getBufferResultFuture(token, getTracker);
+    }
+
+    private ListenableFuture<BufferResult> getBufferResultFuture(long token, GetTracker getTracker)
+    {
+        long maxBytes = getTracker.getMaxSize().toBytes();
+        ImmutableList.Builder<SerializedPage> result = ImmutableList.builder();
```

Comment:
I think we should remove `result`, and just pass these values along through future callbacks.